AssertionError: path is not str type on 'Save permissions'

Hi Germar,

on current git master, my ssh encrypted backup fails on "save permissions" with the following debug output:

```
IINFO: [common/snapshots.py:999 Snapshots._take_snapshot] Save permissions
DEBUG: [common/sshtools.py:258 SSH.unlock_ssh_agent] Private key /home/akki/.ssh/### is already unlocked in ssh agent
DEBUG: [common/encfstools.py:647 Decode.start_process] start 'encfsctl decode' process
DEBUG: [common/encfstools.py:651 Decode.start_process] Call command: encfsctl decode --extpass=backintime-askpass /home/akki/.local/share/backintime/mnt/6B0CA066/mountpoint
Traceback (most recent call last):
  File "/usr/share/backintime/common/backintime.py", line 987, in <module>
    start_app()
  File "/usr/share/backintime/common/backintime.py", line 465, in start_app
    args.func(args)
  File "/usr/share/backintime/common/backintime.py", line 668, in backup
    ret = take_snapshot(cfg, force)
  File "/usr/share/backintime/common/backintime.py", line 77, in take_snapshot
    ret = snapshots.Snapshots( cfg ).take_snapshot( force )
  File "/usr/share/backintime/common/snapshots.py", line 704, in take_snapshot
    ret_val, ret_error = self._take_snapshot( sid, now, include_folders )
  File "/usr/share/backintime/common/snapshots.py", line 1039, in _take_snapshot
    new_snapshot.fileInfo = fileInfoDict
  File "/usr/share/backintime/common/snapshots.py", line 2151, in fileInfo
    assert isinstance(path, str), 'path is not str type: {}'.format(path)
AssertionError: path is not str type: b'/home/akki/Dokumente/Archiv/Korrespondenz'
DEBUG: [common/encfstools.py:761 Decode.close] stop 'encfsctl decode' process
DEBUG: [common/encfstools.py:519 Encode.close] stop 'encfsctl encode' process

```

The path or file name in the last but two line varies with each attempt.
 
This error occurs since commit https://github.com/bit-team/backintime/commit/fa4508ea305978c16193579682653d1900f81000 ; earlier versions work fine...
 
Please let me know if you need any further information...

--- a/common/snapshots.py
+++ b/common/snapshots.py
@@ -811,7 +811,7 @@ class Snapshots:
         assert isinstance(path, bytes), 'path is not bytes type: %s' % path
         if path and os.path.exists(path):
             info = os.stat(path)
-            mode = str(info.st_mode).encode('utf-8', 'replace')
+            mode = info.st_mode
             user = self.get_user_name(info.st_uid).encode('utf-8', 'replace')
             group = self.get_group_name(info.st_gid).encode('utf-8', 'replace')
             fileinfo[path] = (mode, user, group)
@@ -1000,7 +1000,7 @@ class Snapshots:
             self.set_take_snapshot_message( 0, _('Save permission ...') )
 
             permission_done = False
-            fileInfoDict = {}
+            fileInfoDict = FileInfoDict()
             if self.config.get_snapshots_mode() in ['ssh', 'ssh_encfs']:
                 path_to_explore_ssh = new_snapshot.pathBackup(use_mode = ['ssh', 'ssh_encfs']).rstrip( '/' )
                 cmd = self.cmd_ssh(['find', path_to_explore_ssh, '-print'])
@@ -1749,6 +1749,19 @@ class Snapshots:
 
         return (' '.join(items1), ' '.join(items2))
 
+class FileInfoDict(dict):
+    """
+    A `dict` that maps a path (as `bytes`) to (`int`, `bytes`, `bytes`).
+    """
+    def __setitem__(self, key, value):
+        assert isinstance(key, bytes), "key '{}' is not bytes instance".format(key)
+        assert isinstance(value, tuple), "value '{}' is not tuple instance".format(value)
+        assert len(value) == 3, "value '{}' does not have 3 items".format(value)
+        assert isinstance(value[0], int), "first value '{}' is not int instance".format(value[0])
+        assert isinstance(value[1], bytes), "second value '{}' is not bytes instance".format(value[1])
+        assert isinstance(value[2], bytes), "third value '{}' is not bytes instance".format(value[2])
+        super(FileInfoDict, self).__setitem__(key, value)
+
 class SID(object):
     """
     Snapshot ID object used to gather all information for a snapshot
@@ -2161,17 +2174,13 @@ class SID(object):
 
     @fileInfo.setter
     def fileInfo(self, d):
-        assert isinstance(d, dict), 'd is not dict type: {}'.format(d)
+        assert isinstance(d, FileInfoDict), 'd is not FileInfoDict type: {}'.format(d)
         with bz2.BZ2File(self.path(self.FILEINFO), 'wb') as f:
             for path, info in d.items():
-                assert isinstance(path, str), 'path is not str type: {}'.format(path)
-                assert isinstance(info[1], str), 'user is not str type: {}'.format(info[1])
-                assert isinstance(info[2], str), 'group is not str type: {}'.format(info[2])
-                assert isinstance(info[0], int), 'permission is not int type: {}'.format(info[0])
                 f.write(b' '.join((str(info[0]).encode('utf-8', 'replace'),
-                                   info[1].encode('utf-8', 'replace'),
-                                   info[2].encode('utf-8', 'replace'),
-                                   path.encode('utf-8', 'replace') ))
+                                   info[1],
+                                   info[2],
+                                   path))
                                    + b'\n')
 
     #TODO: add arguments 'mode' and 'decode'

--- a/common/test/test_snapshots.py
+++ b/common/test/test_snapshots.py
@@ -700,9 +700,9 @@ class TestSID(GenericSnapshotsTestCase):
                                 '20151219-010324-123',
                                 'fileinfo.bz2')
 
-        d = {}
-        d['/tmp']     = (123, 'foo', 'bar')
-        d['/tmp/foo'] = (456, 'asdf', 'qwer')
+        d = snapshots.FileInfoDict()
+        d[b'/tmp']     = (123, b'foo', b'bar')
+        d[b'/tmp/foo'] = (456, b'asdf', b'qwer')
         sid1.fileInfo = d
 
         self.assertTrue(os.path.isfile(infoFile))

